src/pd/leader.rs

```diff
@@ -1,51 +1,222 @@
 use fxhash::FxHashSet as HashSet;
 use std::result;
-use std::sync::Arc;
-use std::sync::RwLock;
+use std::sync::{Arc, RwLock};
+use std::thread::{self, JoinHandle};
 use std::time::Duration;
 use std::time::Instant;
 
 use futures::future::{loop_fn, ok, Loop};
 use futures::sync::mpsc::{unbounded, UnboundedReceiver, UnboundedSender};
 use futures::sync::oneshot;
-use futures::task::Task;
-use futures::Future;
-use grpc::{CallOption, ChannelBuilder, ClientDuplexReceiver, ClientDuplexSender, Environment};
+use futures::{Future, Sink, Stream};
+use grpc::{CallOption, ChannelBuilder, Environment, WriteFlags};
 use kvproto::pdpb::{
     GetMembersRequest, GetMembersResponse, ResponseHeader, TsoRequest, TsoResponse,
 };
 use kvproto::pdpb_grpc::PdClient;
 use tokio_timer::timer::Handle;
 
-use super::{Error, PdFuture, PdTimestamp, Result, REQUEST_TIMEOUT};
+use super::{Error, PdFuture, PdTimestamp, Result, PD_REQUEST_HISTOGRAM_VEC, REQUEST_TIMEOUT};
+use tokio_core::reactor::{Core, Handle as OtherHandle};
 use util::security::SecurityManager;
+use util::time::duration_to_sec;
 use util::timer::GLOBAL_TIMER_HANDLE;
-use util::{Either, HandyRwLock};
-
-macro_rules! box_err {
-    ($e:expr) => ({
-        use std::error::Error;
-        let e: Box<Error + Sync + Send> = format!("[{}:{}]: {}", file!(), line!(),  $e).into();
-        e.into()
-    });
-    ($f:tt, $($arg:expr),+) => ({
-        box_err!(format!($f, $($arg),+))
-    });
+use util::HandyRwLock;
+
+macro_rules! request {
+    ($cluster_id:expr, $type:ty) => {{
+        let mut request = <$type>::new();
+        let mut header = ::kvproto::pdpb::RequestHeader::new();
+        header.set_cluster_id($cluster_id);
+        request.set_header(header);
+        request
+    }};
+}
+
+type TsoChannel = oneshot::Sender<PdTimestamp>;
+
+type PdRequest = Box<Future<Item = (), Error = ()> + Send>;
+
+pub enum PdTask {
+    TsoInit,
+    TsoRequest,
+    TsoResponse(Vec<oneshot::Sender<PdTimestamp>>, TsoResponse),
+    Request(PdRequest),
+}
+
+struct PdReactor {
+    task_tx: Option<UnboundedSender<Option<PdTask>>>,
```

Comment:
Unbounded channels make me very nervous. Is it possible to opt for a high bound instead? You can have some constant that is a default bound, later if we find some problems we can make it bigger.

Maybe something like 4096 is a generous bound?

Reply:
Tried to change it to bounded channel and realized a problem that it is often used inside reactor thread and blocking behavior might cause deadlock. So sadly only those channels that sending side is used from external thread can be changed to bounded. Some channels has to remain unbounded for this particular reason.

Reply:
Just go over the code again, looks like all channels are used by reactor thread. Looks like it is not possible to have bounded channel.

Reply:
Hm, okay. Seems like we can't change it. Thank you for investigating.